### ISSUE TITLE
zig cc -O broken on MacOS (at least)

### ISSUE THREAD

**Issue by drfuchs** [ziglang/zig#9726]
Doing "zig cc -O3" on MacOS gives a "warning(zld): unknown filetype for positional input file" error.
```
~/Desktop/zig % uname -mprs
Darwin 20.6.0 x86_64 i386

~/Desktop/zig % cat hello.c
#include <stdio.h>
int main(int arc, char** argv)
{
  printf("Hello from %s\n", argv[0]);
  return 0;
}

~/Desktop/zig % cc -O3 hello.c
~/Desktop/zig % ./a.out
Hello from ./a.out

~/Desktop/zig % zig cc hello.c   
~/Desktop/zig % ./a.out
Hello from ./a.out

~/Desktop/zig % zig cc -O3 hello.c
warning(zld): unknown filetype for positional input file: '~/.cache/zig/o/347d130b675906a00127e178567315fb/hello.o'
error: MissingMainEntrypoint

~/Desktop/zig % rm -rf ~/.cache/zig
~/Desktop/zig % zig cc -O3 hello.c          
warning(zld): unknown filetype for positional input file: '~/.cache/zig/o/347d130b675906a00127e178567315fb/hello.o'
error: MissingMainEntrypoint
```

**Comment by floooh** [comment]
Just FYI, I can't reproduce this with the a Zig HEAD version from yesterday (12-Sep-2021) installed with ```brew install zig --HEAD``` on macOS 11.5.

```zig cc -O3 hello.c``` finished without messages and produces a working executable.

Running nm on the produced hello.o file shows the expected internal and external symbols:

```
e38bbc7ed375746b7720fd5852ada131 ➤ nm -v hello.o
                 U _printf
0000000000000000 T _main
```

**Comment by floooh** [comment]
Maybe this and https://github.com/ziglang/zig/issues/9747 are similar to the mysterious problems in older Zig versions installed via homebrew which were caused by Zig being linked dynamically with LLVM instead of statically? In the homebrew recipe this was fixed back in May by this PR by @kubkon: https://github.com/Homebrew/homebrew-core/pull/78336. Just a guess of course.

**Comment by drfuchs** [comment]
Sorry I failed to mention, but my results were indeed from a newly brew-installed zig 0.8.1 (which is the latest available there).
```
~/Desktop/zig % zig version
0.8.1
```

**Comment by kubkon** [comment]
Sorry about this @drfuchs. As @floooh rightly pointed out, I've landed fixes to this and #9747 however they haven't made it into 0.8.1, so currently your best option is to use a nightly built of the compiler or wait until 0.9 release.

**Comment by drfuchs** [comment]
Yup, `brew install zig --HEAD` addressed this one, too, thanks.
